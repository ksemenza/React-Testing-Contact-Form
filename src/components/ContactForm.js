import React, { useState } from "react";
import { useForm } from "react-hook-form";

const ContactForm = () => {
  const [data, setData] = useState();
  const { register, errors, handleSubmit, reset } = useForm({
    mode: "onBlur"
  });
  const onSubmit = data => {
    setData(data);
  };

  return (
    <div className="App">
      <form data-testid='contact-form'  onSubmit={handleSubmit(onSubmit)}>
        <div>
          <label htmlFor="firstName">First Name*</label>
          <input data-testid='fname'
            name="firstName"
            placeholder="bill"
            ref={register({ required: true, maxLength: 20 })}
          />
          {errors.firstName && (
            <p>Looks like there was an error: {errors.firstName.type}</p>
          )}
        </div>

        <div>
          <label htmlFor="lastName">Last Name*</label>
          <input  data-testid='lname'
            name="lastName"
            placeholder="luo"
            ref={register({ required: true })}
          />
          {errors.lastName && (
            <p>Looks like there was an error: {errors.lastName.type}</p>
          )}
        </div>

        <div>
          <label htmlFor="email" placeholder="bluebill1049@hotmail.com">
            Email*
          </label>
          <input  data-testid='email'
          name="email" ref={register({ required: true })} />
          {errors.email && (
            <p>Looks like there was an error: {errors.email.type}</p>
          )}
        </div>
        <div>
          <label htmlFor="message">Message</label>
          <textarea  data-testid='message'
           name="message" ref={register({ required: false })} />
        </div>
        {data && (
          <pre data-testid='submit-form' style={{ textAlign: "left", color: "white" }}>
            {JSON.stringify(data, null, 2)}
          </pre>
        )}
        <input data-testid='submit-btn' type="submit" />
      </form>

     
   
    </div>
  );
};

export default ContactForm;
