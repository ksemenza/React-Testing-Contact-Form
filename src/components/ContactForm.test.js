import React from 'react'
import {render, getByLabelText, getByTestId, getByDisplayValue, fireEvent, getByText, wait, queryByTestId, getByPlaceholderText} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { mount } from 'enzyme';
import enzymeConfig from '../enzymeConfig'
import ContactForm from './ContactForm'
import 'mutationobserver-shim'


test('Required Input', () => {
 const {getByTestId} = render(<ContactForm/>);
 const firstName =getByTestId('fname')
expect(firstName).toBeInTheDocument()

  });

test('Forms Value Inputs', () => {
    const {getByTestId} = render(<ContactForm/>)
    const formCheck = getByTestId('contact-form')

    expect(formCheck).toHaveFormValues({
        firstName:'',
        lastName:'',
        email:''
    })
})

test('Button Specs', () => {
    const {getByTestId} = render(<ContactForm/>)
    const submitBtn = getByTestId('submit-btn')

    expect(submitBtn).toHaveAttribute('type', 'submit')
    expect(submitBtn).not.toHaveAttribute('type', 'reset')
    expect(submitBtn).not.toHaveAttribute('disabled')
})



  test('clicking on button trigger focus on input', () => {
    const { getByPlaceholderText, getByTestId } = render(<ContactForm />)
    const fnameInput = getByTestId('fname')
    const lnameInput = getByTestId('lname')
    fireEvent.click(getByTestId('fname'));
    fireEvent.click(getByTestId('lname'));
    expect(fnameInput).toBe(getByPlaceholderText('bill'));
    expect(lnameInput).toBe(getByPlaceholderText('luo'));
  });

  it('Should capture firstname correctly onSubmit', function() {
      const component = mount(<ContactForm/>)
      const input = component.find('input').at(0);

      input.instance().value='hello'
      input.simulate('change');
    
      
  })