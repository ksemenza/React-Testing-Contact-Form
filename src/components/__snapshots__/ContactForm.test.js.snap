// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusInput matches snapshot 1`] = `
<div
  class="App"
>
  <form
    data-testid="contact-form"
  >
    <div>
      <label
        for="firstName"
      >
        First Name*
      </label>
      <input
        data-testid="fname"
        name="firstName"
        placeholder="bill"
      />
    </div>
    <div>
      <label
        for="lastName"
      >
        Last Name*
      </label>
      <input
        data-testid="lname"
        name="lastName"
        placeholder="luo"
      />
    </div>
    <div>
      <label
        for="email"
        placeholder="bluebill1049@hotmail.com"
      >
        Email*
      </label>
      <input
        data-testid="email"
        name="email"
      />
    </div>
    <div>
      <label
        for="message"
      >
        Message
      </label>
      <textarea
        data-testid="message"
        name="message"
      />
    </div>
    <input
      data-testid="submit-btn"
      type="submit"
    />
  </form>
</div>
`;
